nomad/state/state_store.go

```diff
@@ -5333,6 +5333,110 @@ func (s *StateStore) BootstrapACLTokens(msgType structs.MessageType, index uint6
 	return txn.Commit()
 }
 
+// UpsertOneTimeToken is used to create or update a set of ACL
+// tokens. Validating that we're not upserting an already-expired token is
+// made the responsibility of the caller to facilitate testing.
+func (s *StateStore) UpsertOneTimeToken(msgType structs.MessageType, index uint64, token *structs.OneTimeToken) error {
+	txn := s.db.WriteTxnMsgT(msgType, index)
+	defer txn.Abort()
+
+	// Check if a one-time token already exists for this ACL token
+	existing, err := txn.First("one_time_token", "id", token.AccessorID)
+	if err != nil {
+		return fmt.Errorf("token lookup failed: %v", err)
+	}
+
+	// we expect the RPC call to set the ExpiresAt but if it's somehow not
+	// set, set it now.
+	if token.ExpiresAt.IsZero() {
+		time.Now().Add(time.Minute * 10)
+	}
+
+	// Update all the indexes
+	if existing != nil {
+		existTK := existing.(*structs.OneTimeToken)
+		token.CreateIndex = existTK.CreateIndex
```

Comment:
Given that we're overwriting the token entirely, I could see an argument for not bothering to do this and just make a new `CreateIndex` that's the same as the `ModifyIndex`?